atmos_model.F90

```diff
@@ -1926,683 +2302,659 @@ subroutine assign_importdata(rc)
-    integer                :: j, i, ix, nb, isc, iec, jsc, jec, idx
+    integer                :: j, i, k, ix, nb, nk, isc, iec, jsc, jec, idx
+    integer                :: sphum, liq_wat, ice_wat, o3mr
     real(GFS_kind_phys)    :: rtime, rtimek
-!
-    if (mpp_pe() == mpp_root_pe()) print *,'enter setup_exportdata'
 
-    isc = GFS_control%isc
-    iec = GFS_control%isc+GFS_control%nx-1
-    jsc = GFS_control%jsc
-    jec = GFS_control%jsc+GFS_control%ny-1
+    integer                                     :: localrc, rc
+    integer                                     :: n,dimCount
+    logical                                     :: isCreated
+    type(ESMF_TypeKind_Flag)                    :: datatype
+    character(len=ESMF_MAXSTR)                  :: fieldName
+    real(kind=ESMF_KIND_R4), dimension(:,:), pointer   :: datar42d
+    real(kind=ESMF_KIND_R8), dimension(:,:), pointer   :: datar82d
+    real(kind=ESMF_KIND_R8), dimension(:,:,:), pointer :: datar83d
+
+    isc = Atm_block%isc
+    iec = Atm_block%iec
+    jsc = Atm_block%jsc
+    jec = Atm_block%jec
+    nk  = Atm_block%npz
 
     rtime  = one / GFS_control%dtp
     rtimek = GFS_control%rho_h2o * rtime
-!    print *,'in cplExp,dim=',isc,iec,jsc,jec,'nExportFields=',nExportFields
-!    print *,'in cplExp,GFS_data, size', size(GFS_data)
-!    print *,'in cplExp,u10micpl, size', size(GFS_data(1)%coupling%u10mi_cpl)
 
-    if(.not.allocated(exportData)) then
-      allocate(exportData(isc:iec,jsc:jec,nExportFields))
-    endif
+    do n=1, size(exportFields)
+
+      datar42d => null()
+      datar82d => null()
+      datar83d => null()
+
+      isCreated = ESMF_FieldIsCreated(exportFields(n), rc=localrc)
+      if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__, rcToReturn=rc)) return
+      if (.not. isCreated) cycle
+
+      call ESMF_FieldGet(exportFields(n), name=fieldname, dimCount=dimCount, typekind=datatype, rc=localrc)
+      if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__, rcToReturn=rc)) return
+      if (datatype == ESMF_TYPEKIND_R8) then
+         if (dimCount == 2) then
+           call ESMF_FieldGet(exportFields(n),farrayPtr=datar82d,localDE=0, rc=localrc)
+           if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__, rcToReturn=rc)) return
+         else if (dimCount == 3) then
+           call ESMF_FieldGet(exportFields(n),farrayPtr=datar83d,localDE=0, rc=localrc)
+           if (ESMF_LogFoundError(rcToCheck=localrc, msg=ESMF_LOGERR_PASSTHRU, line=__LINE__, file=__FILE__, rcToReturn=rc)) return
+         else
+            write(0,*)'not implemented dimCount ',dimCount, trim(fieldname)
+            call ESMF_Finalize(endflag=ESMF_END_ABORT)
+         endif
+      else if (datatype == ESMF_TYPEKIND_R4) then
+         if (dimCount == 2) then
+           call ESMF_FieldGet(exportFields(n),farrayPtr=datar42d,localDE=0, rc=localrc)
```

Comment:
I see only datar83d is used later when setting up the pointers, I am not sure if this will work for 32BIT=Y, for which we may need to set up datar43d. It may not be an issue for fv3-mom6-cice6-wav coupling as those coupled fields are from physics which is always running with real8. We can update this later.